core/app/models/spree/tax_rate.rb

```diff
@@ -16,51 +6,31 @@ class TaxRate < Spree::Base
     include Spree::AdjustmentSource
 
     belongs_to :zone, class_name: "Spree::Zone", inverse_of: :tax_rates
-    belongs_to :tax_category, class_name: "Spree::TaxCategory", inverse_of: :tax_rates
+    belongs_to :tax_category,
+      class_name: "Spree::TaxCategory",
+      inverse_of: :tax_rates
 
     validates :amount, presence: true, numericality: true
     validates :tax_category_id, presence: true
-    validates_with DefaultTaxZoneValidator
 
-    scope :by_zone, ->(zone) { where(zone_id: zone) }
-
-    def self.potential_rates_for_zone(zone)
-      select("spree_tax_rates.*, spree_zones.default_tax").
-        joins(:zone).
-        merge(Spree::Zone.potential_matching_zones(zone)).
-        order("spree_zones.default_tax DESC")
-    end
+    scope :by_zone, -> (zone) { where(zone_id: zone.id) }
+    scope :potential_rates_for_zone,
+      -> (zone) do
+        where(zone_id: Spree::Zone.potential_matching_zones(zone).pluck(:id))
```

Comment:
It replaces the method deleted above (which I think would cause the same n+1 if so). It is only called from within tax_rate.rb (is there such a thing as private scopes?), in order to match the zone of the tax rate with the zone of the order. This happens often: 
- Every time the order is updated
- Every time a price has to be displayed for an order with a tax zone outside the default zone.